### ISSUE TITLE
:recycle: Refactor handling of gate matrices and inverses

### ISSUE THREAD

**Issue by burgholzer** [cda-tum/mqt-core#484]
After #460, all gates in MQT Core have a proper inverse now.
A natural follow-up would be to consolidate the following methods:
- https://github.com/cda-tum/mqt-core/blob/95541a64ee6174bb78baf2dc76f412018a270780/src/operations/StandardOperation.cpp#L471-L559
- https://github.com/cda-tum/mqt-core/blob/95541a64ee6174bb78baf2dc76f412018a270780/include/dd/Operations.hpp#L27-L93
- https://github.com/cda-tum/mqt-core/blob/95541a64ee6174bb78baf2dc76f412018a270780/include/dd/Operations.hpp#L117-L237

There is lots of redundancy in the handling of the `inverse` throughout the code base.
I do not have a perfect solution for resolving that in mind. It would be pretty convenient though if any operation could return its gate matrix. So essentially to refactor https://github.com/cda-tum/mqt-core/blob/main/include/dd/GateMatrixDefinitions.hpp so that the appropriate matrix can be queried from an `Operation`. 
To this end, it could be nice to adopt a strategy similar to Qiskit where there are individual classes for gates that all derive from `StandardOperation` (and potentially `SymbolicOperation` for parametrized gates). As an example:
- there would be an `XGate` class that provides a method to get the single-qubit `GateMatrix`, the type of the gate (already covered by StandardOperation), the inverse type, the inverse GateMatrix and potentially other useful methods.
- The challenge will be to get the interface right so that there is a standardized way to call the newly introduced methods. In that regard, further compile-time/`constexpr` convenience functions like `isSingleTargetGate`, `isTwoTargetGate`, and `isThreeOrMoreTargetGate` would probably make sense.
- Naturally, this creates quite some new classes, but I believe it makes the library structure a little simpler. Needs a little experimenting on the refactor though.
- Maybe the above actually makes some of the operations (such as `iSWAPdg` redundant again.

_Originally posted by @burgholzer in https://github.com/cda-tum/mqt-core/pull/460#pullrequestreview-1738494988_
